### PR TITLE
ENH: Add ability to override which checkpoint to use for inference

### PR DESCRIPTION
**Pull request by ant0nsc**
Models that use a custom logic for choosing best checkpoints can't presently use those for inference. This PR adds a method to the container to control that.

### CODE REVIEW

File: hi-ml/src/health_ml/lightning_container.py [hunk 1]
```diff
@@ -135,6 +136,16 @@ def before_training_on_all_ranks(self) -> None:
         """
         pass
 
+    def get_checkpoint_to_test(self) -> Path:
+        """Returns the path of the model checkpoint that should be used for testing/inference. By default, this will
+        return the checkpoint that is written in the last training epoch. Override this method if you implement a
+        custom checkpointing logic, for example if you added a model checkpoint callback that looks at validation
+        accuracuy.
```
Comment (harshita-s) [inline hi-ml/src/health_ml/lightning_container.py:143]: small typo in 'accuracy'

File: hi-ml/src/health_ml/utils/checkpoint_handler.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        if is_global_rank_zero():
+            checkpoints = list(self.container.checkpoint_folder.rglob("*"))
+            logging.info(f"Available checkpoints: {len(checkpoints)}")
+            for f in checkpoints:
+                logging.info(f)
+        return find_recovery_checkpoint_on_disk_or_cloud(self.container.checkpoint_folder)
+
+    def get_checkpoint_to_test(self) -> Path:
+        """
+        Find the model checkpoint that should be used for inference. If a run recovery is provided, or if the model
```
Comment (harshita-s) [inline hi-ml/src/health_ml/utils/checkpoint_handler.py:73]: Checking if the run recovery case is included somewhere (earlier we had a case with recovery checkpoints `if self.run_recovery or self.has_continued_training`) If this isn't required now, perhaps the docs need to be updated?

Comment (ant0nsc) [reply hi-ml/src/health_ml/utils/checkpoint_handler.py:73]: Thanks, updated.

File: hi-ml/src/health_ml/run_ml.py [hunk 3]
```diff
@@ -175,15 +169,12 @@ def is_crossval_disabled_or_child_0(self) -> bool:
             return self.container.crossval_index == 0
         return True
 
-    def run_inference(self, checkpoint_paths: List[Path]) -> None:
+    def run_inference(self, checkpoint_path: Path) -> None:
         """
         Run inference on the test set for all models.
 
         :param checkpoint_paths: The path to the checkpoint that should be used for inference.
```
Comment (mebristo) [inline hi-ml/src/health_ml/run_ml.py:176]: Param name needs updating to checkpoint_path (singular)

File: hi-ml/src/health_ml/deep_learning_config.py [hunk 4]
```diff
@@ -128,15 +127,11 @@ class WorkflowParams(param.Parameterized):
     This class contains all parameters that affect how the whole training and testing workflow is executed.
     """
     random_seed: int = param.Integer(42, doc="The seed to use for all random number generators.")
-    weights_url: List[str] = param.List(default=[], class_=str,
-                                        doc="If provided, a set of urls from which checkpoints will be downloaded"
-                                            "and used for inference.")
-    local_weights_path: List[Path] = param.List(default=[], class_=Path,
-                                                doc="A list of checkpoints paths to use for inference, "
-                                                    "when the job is running outside Azure.")
-    model_id: str = param.String(default="",
```
Comment (mebristo) [inline hi-ml/src/health_ml/deep_learning_config.py:137]: Why do we get rid of model_id in this PR?

File: hi-ml/src/health_ml/deep_learning_config.py [hunk 5]
```diff
@@ -20,8 +20,7 @@
 from health_ml.utils.common_utils import (CHECKPOINT_FOLDER,
                                           create_unique_timestamp_id,
                                           DEFAULT_AML_UPLOAD_DIR,
-                                          DEFAULT_LOGS_DIR_NAME,
-                                          parse_model_id_and_version)
```
Comment (mebristo) [inline hi-ml/src/health_ml/deep_learning_config.py:24]: I think this file is the only place we were calling this function? If so, can we delete it now?

File: hi-ml/src/health_ml/deep_learning_config.py (hunk 4 above)
Comment (ant0nsc) [reply hi-ml/src/health_ml/deep_learning_config.py:137]: because it was not used.

**Review by harshita-s — APPROVED** [review]

**Pull request merged**